modules/user/lib/user-sync-vebal-lock-balance.service.ts

```diff
@@ -0,0 +1,113 @@
+import _ from 'lodash';
+import { prismaBulkExecuteOperations } from '../../../prisma/prisma-util';
+import RewardsOnlyGaugeAbi from './abi/RewardsOnlyGauge.json';
+import { Multicaller } from '../../web3/multicaller';
+import { formatFixed } from '@ethersproject/bignumber';
+import { PrismaPoolStakingType } from '@prisma/client';
+import { networkContext } from '../../network/network-context.service';
+import { GaugeSubgraphService } from '../../subgraphs/gauge-subgraph/gauge-subgraph.service';
+import { veBalLocksSubgraphService } from '../../subgraphs/veBal-locks-subgraph/veBal-locks-subgraph.service';
+import { BigNumber } from 'ethers';
+import VeBalABI from '../../vebal/abi/vebal.json';
+import mainnet from '../../../config/mainnet';
+
+export class UserSyncVebalLockBalanceService implements UserStakedBalanceService {
+    get chain() {
+        return networkContext.chain;
+    }
+
+    private readonly veBalPoolId = '0x5c6ee304399dbdb9c8ef030ab642b10820db8f56000200000000000000000014';
+
+    public async initStakedBalances(stakingTypes: PrismaPoolStakingType[]): Promise<void> {
+        if (!stakingTypes.includes('VEBAL') && this.chain !== 'MAINNET') {
+            return;
+        }
+
+        console.log('initStakedVebalBalances: Starting loading users and onchain balances...');
+        await this.syncBalances(true);
+
+        console.log('initStakedVebalBalances: finished...');
+    }
+
+    public async syncChangedStakedBalances(): Promise<void> {
+        await this.syncBalances(false);
+    }
+
+    private async syncBalances(init: boolean): Promise<void> {
+        const subgraphVeBalHolders = await veBalLocksSubgraphService.getAllveBalHolders();
+        const metadata = await veBalLocksSubgraphService.getMetadata();
+
+        let operations: any[] = [];
+        // for mainnet, we get the vebal balance form the vebal contract
+        const multicall = new Multicaller(networkContext.data.multicall, networkContext.provider, VeBalABI);
+
+        let response = {} as {
+            [userAddress: string]: {
+                balance: BigNumber;
+                locked: BigNumber[];
+            };
+        };
+
+        for (const holder of subgraphVeBalHolders) {
+            multicall.call(`${holder.user}.locked`, networkContext.data.veBal!.address, 'locked', [holder.user]);
+
+            // so if we scheduled more than 100 calls, we execute the batch
+            if (multicall.numCalls >= 100) {
+                response = _.merge(response, await multicall.execute());
+            }
+        }
+
+        if (multicall.numCalls > 0) {
+            response = _.merge(response, await multicall.execute());
+        }
+
+        operations.push(
+            prisma.prismaUser.createMany({
+                data: subgraphVeBalHolders.map((holder) => ({ address: holder.user.toLowerCase() })),
+                skipDuplicates: true,
+            }),
+        );
+
+        if (init) {
```

Comment:
What is the purpose of the init function? Is it one time thing? Can we refactor it to remove init branching to a separate function?